airflow/contrib/hooks/sagemaker_hook.py

```diff
@@ -0,0 +1,167 @@
+        bucket, key = self.parse_s3_url(s3url)
+        s3hook = S3Hook(aws_conn_id=self.aws_conn_id)
+        if s3hook.check_for_key(key=key, bucket_name=bucket):
+            raise AirflowException("The input S3 Url {} does not exist ".format(s3url))
+        return True
+
+    @staticmethod
+    def is_key(key, dict, dict_name="configuration"):
+        # type: (string, dict, string) -> bool
+        if key not in dict:
+            raise AirflowException("{} is required, but is missing in the {} provided"
+                                   .format(key, dict_name))
+        return True
+
+    def check_valid_training_input(self, config):
+        InstanceTypes = ['ml.m4.xlarge', 'ml.m4.2xlarge', 'ml.m4.4xlarge',
+                         'ml.m4.10xlarge', 'ml.m4.16xlarge', 'ml.m5.large',
+                         'ml.m5.xlarge', 'ml.m5.2xlarge', 'ml.m5.4xlarge',
+                         'ml.m5.12xlarge', 'ml.m5.24xlarge', 'ml.c4.xlarge',
+                         'ml.c4.2xlarge', 'ml.c4.4xlarge', 'ml.c4.8xlarge',
+                         'ml.p2.xlarge', 'ml.p2.8xlarge', 'ml.p2.16xlarge',
+                         'ml.p3.2xlarge', 'ml.p3.8xlarge', 'ml.p3.16xlarge',
+                         'ml.c5.xlarge', 'ml.c5.2xlarge', 'ml.c5.4xlarge',
+                         'ml.c5.9xlarge', 'ml.c5.18xlarge']
+        self.is_key("TrainingJobName", config)
+        self.is_key("AlgorithmSpecification", config)
+        self.is_key("TrainingImage", config['AlgorithmSpecification'],
+                    dict_name="AlgorithmSpecification")
+        self.is_key("TrainingInputMode", config['AlgorithmSpecification'],
+                    dict_name="AlgorithmSpecification")
+        self.is_key("RoleArn", config)
+        self.is_key("InputDataConfig", config)
+        for channel in config["InputDataConfig"]:
+            self.is_key("ChannelName", channel, dict_name="InputDataConfig")
+            self.is_key("DataSource", channel, dict_name="InputDataConfig")
+            self.is_key("S3DataSource", channel['DataSource'], dict_name="DataSource")
+            self.is_key("S3DataType", channel['DataSource']['S3DataSource'],
+                        dict_name="S3DataSource")
+            self.is_key("S3Uri", channel['DataSource']['S3DataSource'],
+                        dict_name="S3DataSource")
+        self.is_key("OutputDataConfig", config)
+        self.is_key("S3OutputPath", config['OutputDataConfig'],
+                    dict_name="OutputDataConfig")
+        self.is_key("ResourceConfig", config)
+        self.is_key("InstanceType", config['ResourceConfig'], "ResourceConfig")
+        self.is_key("InstanceCount", config['ResourceConfig'], "ResourceConfig")
+        self.is_key("VolumeSizeInGB", config['ResourceConfig'], "ResourceConfig")
+        if self.is_key("VpcConfig", config):
+            self.is_key("SecurityGroupIds", config['VpcConfig'], "VpcConfig")
+            self.is_key("Subnets", config['VpcConfig'], "VpcConfig")
```

Comment:
You could check the notebook example kmeans-lowlevel (https://github.com/awslabs/amazon-sagemaker-examples/blob/master/sagemaker-python-sdk/1P_kmeans_lowlevel/kmeans_mnist_lowlevel.ipynb), not all keys here are required. For example, 'Subnets' and 'SecurityGroupIds' are not provided in the sample notebook.

Reply:
After testing with the API, I found out that boto actually do some preliminary checks, which includes checking the format of input, so there is no longer need to include the checks here. Still investigating what else can be included.